Objective-C/CBLArray.m

```diff
@@ -0,0 +1,342 @@
+- (void) setArray:(NSArray *)array {
+    // Detach all objects that we are listening to for changes:
+    [self detachChildChangeListeners];
+    
+    NSMutableArray* result = [NSMutableArray arrayWithCapacity: [array count]];
+    for (id item in array) {
+        [result addObject: [self prepareValue: item]];
+    }
+    
+    _array = result;
+    [self setChanged];
+}
+
+
+- (void) setObject: (id)value atIndex: (NSUInteger)index {
+    id oldValue = [self objectAtIndex: index];
+    if (!$equal(value, oldValue)) {
+        value = [self prepareValue: value];
+        [self detachChangeListenerForObject: oldValue];
+        [self setValue: value atIndex: index isChange: YES];
+    }
+}
+
+
+- (void) addObject: (id)value  {
+    [_array addObject: [self prepareValue: value]];
+    [self setChanged];
+}
+
+
+- (void) insertObject: (id)value atIndex: (NSUInteger)index {
+    [_array insertObject: [self prepareValue: value] atIndex: index];
+    [self setChanged];
+}
+
+
+- (void) removeObjectAtIndex:(NSUInteger)index {
+    id value = _array[index];
+    [self detachChangeListenerForObject: value];
+    [_array removeObjectAtIndex: index];
+    [self setChanged];
+}
+
+
+#pragma mark - SUBSCRIPTION
+
+
+- (CBLFragment*) objectAtIndexedSubscript: (NSUInteger)index {
+    id value = index < self.count ? [self objectAtIndex: index] : nil;
+    return [[CBLFragment alloc] initWithValue: value parent: self parentKey: @(index)];
+}
+
+
+#pragma mark - CHANGE LISTENER
+
+
+- (void) addChangeListener: (id<CBLObjectChangeListener>)listener {
+    if (!_changeListeners)
+        _changeListeners = [NSMapTable weakToStrongObjectsMapTable];
+    NSInteger count = [[_changeListeners objectForKey: listener] integerValue] + 1;
+    [_changeListeners setObject: @(count) forKey: listener];
+}
+
+
+- (void) removeChangeListener: (id<CBLObjectChangeListener>)listener {
+    NSInteger count = [[_changeListeners objectForKey: listener] integerValue] - 1;
+    if (count > 0)
+        [_changeListeners setObject: @(count) forKey: listener];
+    else
+        [_changeListeners removeObjectForKey: listener];
+}
+
+
+- (void) detachChildChangeListeners {
+    for (id object in _array) {
+        [self detachChangeListenerForObject: object];
+    }
+}
+
+
+- (void) detachChangeListenerForObject: (id)object {
+    if ([object isKindOfClass: [CBLSubdocument class]]) {
+        CBLSubdocument* subdocument = (CBLSubdocument*)object;
+        [subdocument.dictionary removeChangeListener: self];
+    } else if ([object isKindOfClass: [CBLArray class]]) {
+        CBLArray* array = (CBLArray*)object;
+        [array removeChangeListener: self];
+    }
+}
+
+- (void) notifyChangeListeners {
+    for (id <CBLObjectChangeListener> listener in _changeListeners) {
+        [listener objectDidChange: self];
+    }
+}
+
+
+#pragma mark - CHANGE LISTENING
+
+
+- (void) objectDidChange: (id)object {
+    [self setChanged];
+}
+
+
+#pragma mark - PRIVATE
+
+
+- (void) loadBackingData {
```

Comment:
It loading all data from fleece into an (mutable) array for edit. Probably I should change the method name.